### DISCUSSION TITLE
ERT Field

### DISCUSSION THREAD

**Discussion by GeorgeLiang3** [gimli-org/pyGIMLi#300] [Q&A]
I tried to use a 'dd' scheme in a 3D model.
`scheme = ert.createERTData(elecs=np.linspace(start=-15, stop=15, num=4),
                           schemeName='dd')`
so the 4 electrodes are placed evenly at 
`([[-15.,   0.,   0.],
       [ -5.,   0.,   0.],
       [  5.,   0.,   0.],
       [ 15.,   0.,   0.]])`
and evaluate the potential field 
`field_het = simulateERT(mesh, res=res, scheme=scheme, sr=False,calcOnly=True, verbose=True,returnFields=returnFields)`
I was expecting one field to be returned, as we have specified a,b,m,n. However, the 3 fields with different pair of injection current electrodes were returned? Could someone help me understand this please?

Cheers,
George

**Comment by halbmy** [comment]
The resulting scheme contains four electrodes but only one measurement (A=1, B=2, M=3, N=4).

When simulating ERT, we typically model single electrodes (i.e. four potential distributions) and superpose them for later use. By specifying `return_fields=True`, one obtains a potential matrix with four rows (`field_het.rows()` yields 4) and the number of columns (`field_het.cols()`) is the number of nodes in the mesh. So you could use `field_het[1]-field_het[0]` to see the dipole field from AB=1-2. Note that the actual data scheme (here DD) is not really important with `return_fields`.

**Reply by GeorgeLiang3** [reply]
@halbmy  Thank you, but the `field_het.rows()` is 3 isn't it? sorry if I was wrong

```
scheme = ert.createERTData(elecs=np.linspace(start=-15, stop=15, num=4),
                           schemeName='dd')
print(scheme.sensorPositions().array())
```
scheme sensor positions with 4 electrodes are

```
[[-15.   0.   0.]
 [ -5.   0.   0.]
 [  5.   0.   0.]
 [ 15.   0.   0.]]
```
```
field_het = simulateERT(mesh, res=res, scheme=scheme, sr=False,calcOnly=True, verbose=True,returnFields=True)
print(field_het.rows())
```
output is 
```
3
```

**Reply by halbmy** [reply]
No, the output of `het.rows()` should be 4 as long as all four electrodes could be found in the mesh. I used a very simple example (calls might be slightly different in my developer version)
```
import numpy as np
import pygimli as pg
import pygimli.meshtools as mt
from pygimli.physics import ert

scheme = ert.createERTData(elecs=np.linspace(start=-15, stop=15, num=4), schemeName='dd')
mesh = mt.createParaMesh(scheme)
print(mesh)
field = ert.simulate(mesh, res=1.0, scheme=scheme, sr=False, calcOnly=True, verbose=True, returnFields=True)
print(field.rows(), field.cols())
```
obtaining `4 78` (of course the mesh does not make any sense for modelling).

**Reply by GeorgeLiang3** [reply]
Okay, I see, the problem comes in the 3D case, I was trying to use the scheme directly in 3D mesh
```

plc = mt.createCube(size=[30, 30, 10], pos=[0, 0, -15], boundaryMarker=1)

plc = createPLC()
mesh = mt.createMesh(plc)

field = ert.simulate(mesh, res=1.0, scheme=scheme, sr=False, calcOnly=True, verbose=True, returnFields=True)
print(field.rows(), field.cols())
```
Would you please point me some examples where I can see how to generate the 3D scheme, please. I hardly find any example with a 3D setup on. The only one I found on the website is [Example](https://www.pygimli.org/_examples_auto/3_dc_and_ip/plot_modTank3d.html#sphx-glr-examples-auto-3-dc-and-ip-plot-modtank3d-py)

**Reply by halbmy** [reply]
Make sure your electrodes are inside the mesh (and also as nodes in the mesh!) so you will end up in 4.

For other things like data schemes, you might also have a look at the BERT repository at https://gitlab.com/resistivity-net/bert

**Reply by GeorgeLiang3** [reply]
Thanks @halbmy , sorry for the dummy code, I find out it was the reference nodes causing the problem, if no reference points were added, 3 fields will be returned
```
plc.createNode([0.5, 0.5, -0.5], marker=-999)
plc.createNode([0.75, 0.25, 0.5], marker=-1000)
```

**Reply by halbmy** [reply]
Yes, thats true. In case of a Neumann body, you need a current reference (-999) against which the single electrode currents are injected so that the divergence of the whole model domain is zero. If there is no such node, the last electrode is taken. This influences the potential fields (`returnFields=True`) but not the data itself.
